### ISSUE TITLE
Custom tensor attributes not preserved with registered functions

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#110461]
In the following snippet
```.py
import torch

class Fn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        return x/x.coef

    @staticmethod
    def backward(ctx):
        pass


def my_fn_impl(x):
    x.coef = 2.
    print("COEF", x.coef)
    return Fn.apply(x)

from torch.library import Library

ns = 'capturable'
lib = Library(ns, "DEF")
fwd_name = lib.define(f"capturable::myop(Tensor x) -> Tensor")
lib.impl(fwd_name, my_fn_impl, 'CPU')

x = torch.randn(2, 3)
print(my_fn_impl(x))  #fine
print(torch.ops.capturable.myop(x)) # COEF is printed, but autograd fn says Tensor has no attribute coef
```
when `my_fn_impl` is registered on CPU key, an attribute coef that's attached to tensor in my_fn_impl is stripped by Autograd function
According to @zou3519 
>For some reason, HermeticPyObjectTLS is set, which disables PyObject preservation (the thing that would make .coef show up on the Tensor)

**Comment by zou3519** [comment]
It looks like torch.library uses PythonKernelHolder that sets HermeticPyObjectTLS to True.
https://github.com/pytorch/pytorch/blob/a8a31bc16589ac150d1bf90decf6d52c453aa07c/torch/csrc/utils/python_dispatch.cpp#L174

it seems like we do this for torch::deploy/multipy, which registers custom ops to the main interpreter. In that case, calling the operator from a different interpreter shouldn't preserve PyObject because we run the risk of mixing PyObject from different interpreter. Maybe we could carve out an exception for the main interpreter to not set HermeticPyObjectTLS to True? cc @ezyang

**Comment by ezyang** [comment]
Nah, this is intentional. The idea is that every custom operator definition must, in principle, be implementable in pure C++. This isn't the case here, so I would claim we're allowed to lose the attribute.

What do you actually want to do @ngimel?

**Comment by ngimel** [comment]
I want to be able to script or trace `my_fn_impl` as described here https://github.com/pytorch/pytorch/pull/62660#issuecomment-1006071930, hence I'm registering it to the Library. Old code based on #62660 stopped working with pytorch 2.1/python 3.11
